### PR TITLE
Recursive pack

### PR DESCRIPTION
**Pull request by thomasmars**
-r flag for h5p pack that will include all dependencies.

### CODE REVIEW

**Comment by icc** [comment]
I'm a bit concerned about all the refactoring. I do agree that we should use promises, but I'm unsure if we're approaching this the correct way. Refactoring a bit here and there creates a strange mix of everything which easily gets confusing. 

1) Could we have used some sort of wrapper like `Promise.denodeify()` instead of having each function use the Promise constructor? Optionally, each function could have been a promise instead. Something just feels wrong when the whole body of the function is wrapped inside a
```javascript
let doStuff = function (args) { return new Promise(function(res,rej) {
   ...
}); }
```

2) The return type of a function shouldn't need to be a part of the function name. E.g. _processReposPromise_, _getIdNumber_ or _getObjectsArray_. In most cases it's redundant or just confusing.

3) We should use promises for the native node functions we're using as well. 

It feels a bit like we've added a lot of complexity, while we probably should just have refactored the whole _lib/h5p.js_ into classes or something. Testing all of these functions and changes will be quite the job.

**Comment by thomasmars** [comment]
Yeah, there will be quite a bit of redundancy while supporting both callback functions and promise functions, but I think if we are gonna improve the code-base we have to do it gradually, which will create redundancy until all functions have transitioned to using promises, at which point we can remove the old functions. 

1) I'm not familiar with denodify, but I don't see how it would be faster or more readable than doing a gradual transition to promises, I might be wrong here though. There might be a better way to use promises here, but it is quite hard to pick apart the monolith gradually, which results in sub-optimal code until everything has been refactored.

2) I agree. This was meant as a means to distinguish it from its callback counterpart and to keep track of which functions performed the same work, but in different fashions.

3) Agree, but this must be done gradually, we can not hope to fix the whole code-base in one go.

Some complexity has been added, but I also think parts of the code-base has been simplified. The work to refactor h5p.js into classes has been started by extracting commands into separate classes and utility libraries. Everything that has been added has also got accompanying unit tests to ease the test process.

I'm not saying that this whole pull request should be processed in one go, but perhaps there are useful commits that can be cherry-picked and added gradually to the code-base.

**Comment by icc** [comment]
I see… if we leave the code style and structure, for now, I only have two issues with merging this:

1) It's impossible to diff the changes made on the function since they are "cloned". 
To solve this I suggest we try to make the *Promise functions use the old function and then apply the changes to the old function. Since I assume that both functions should do the same except for the Promise functionality? At least, I don't see the need for having the old function doing something different than the new… Or we just remove the old function altogether and only use the promises.

2) The idea behind separation of the CLI and Lib part was to be able to use the Lib in other contexts, e.g. with a graphical, web or network UI. This means that the Lib part shouldn't write any output and just pass along its results to whoever want to present it to the user (in whatever way or form they want).

**Comment by thomasmars** [comment]
1) I'm not sure how feasible this is, but we can try :)

2) There is an output utility library constructed with exactly this idea in mind. Output from libs are passed to the output utility libraries, where it can be processed/presented however we want.

**Comment by icc** [comment]
Please avoid adding things to this pull request that isn't directly related. It will only slow down the review process…

File: lib/commands/check-versions.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
+const Input = require('../utility/input');
+const Output = require('../utility/output');
+const Repository = require('../utility/repository');
+const AdmZip = require('adm-zip');
```
Comment (icc) [inline lib/commands/check-versions.js]: @thomasmars why do we need to use this instead of the already required `archiver`?

**Comment by icc** [comment]
Very thorough job on the recursive pack.

File: lib/commands/check-versions.js (hunk 1 above)
Comment (thomasmars) [inline lib/commands/check-versions.js]: @icc As far as I can see `archiver` only allows you to generate archives, not unzip/open them.

Comment (icc) [inline lib/commands/check-versions.js]: Ah, ok. We should probably try to move away from it then :D

**Pull request merged**